Fix rope_deltas corruption in Qwen2.5VL during CFG generation

notkisk opened the pull request:
## Summary

  Fixes #39749 
 

  This PR addresses a critical issue where the `rope_deltas` attribute in Qwen2.5VL models gets corrupted during Classifier-Free Guidance (CFG) generation due to shared mutable state between forward passes.

  **Problem**: During CFG generation, the model performs two forward passes (conditional and unconditional). The `rope_deltas` state was being modified during the first pass and incorrectly reused in the second pass, leading to position embedding corruption and incorrect generation
  results.

  **Solution**: Implemented proper state management by:
  - Adding `_update_model_kwargs_for_generation` method to preserve `rope_deltas` in model kwargs
  - Modifying `prepare_inputs_for_generation` to restore `rope_deltas` from kwargs when available
  - Ensuring each forward pass gets the correct `rope_deltas` state

  **Impact**: Fixes CFG generation for Qwen2.5VL models, ensuring correct position calculations and generation quality.

  ## Changes Made

  - Modified `src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py` with the fix
  - Regenerated `src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py` from the modular source
  - Added comprehensive state management for `rope_deltas` during generation

  ## Test Plan

  - [x] Verified the fix addresses the original issue reproduction case
  - [x] Tested CFG generation with multiple forward passes
  - [x] Confirmed `rope_deltas` state is properly preserved and restored
  - [x] Ran quality checks and style formatting
  - [x] Ensured backward compatibility with existing functionality

  The fix is ready and properly references issue #39749 as requested. The branch contains all the necessary changes to resolve the rope_deltas corruption issue during CFG generation.

notkisk committed:
Fix rope_deltas state corruption in Qwen2.5VL during CFG generation

This commit addresses an issue where the rope_deltas attribute in Qwen2.5VL
models gets corrupted during Classifier-Free Guidance (CFG) generation due
to shared mutable state between forward passes.

The problem occurred because:
1. CFG generation requires two forward passes per iteration (conditional and unconditional)
2. self.rope_deltas was modified during forward passes (line 643: self.rope_deltas = rope_deltas)
3. The second forward pass used corrupted rope_deltas, leading to incorrect position calculations

Changes made:
- Added _update_model_kwargs_for_generation method to preserve rope_deltas state
- Modified prepare_inputs_for_generation to restore rope_deltas from model kwargs
- Both changes implemented in modular_qwen2_5_vl.py and regenerated to modeling_qwen2_5_vl.py

This ensures that each forward pass in CFG generation has access to the correct
rope_deltas state, preventing position embedding corruption and maintaining
generation quality.

notkisk committed:
Fix rope_deltas state corruption in Qwen2.5VL during CFG generation

This commit addresses an issue where the rope_deltas attribute in Qwen2.5VL
models gets corrupted during Classifier-Free Guidance (CFG) generation due
to shared mutable state between forward passes.

The problem occurred because:
1. CFG generation requires two forward passes per iteration (conditional and unconditional)
2. self.rope_deltas was modified during forward passes (line 643: self.rope_deltas = rope_deltas)
3. The second forward pass used corrupted rope_deltas, leading to incorrect position calculations

Changes made:
- Added _update_model_kwargs_for_generation method to preserve rope_deltas state
- Modified prepare_inputs_for_generation to restore rope_deltas from model kwargs
- Both changes implemented in modular_qwen2_5_vl.py and regenerated to modeling_qwen2_5_vl.py

This ensures that each forward pass in CFG generation has access to the correct
rope_deltas state, preventing position embedding corruption and maintaining
generation quality.

notkisk committed:
Override rope_deltas method in GLM4V to prevent inheritance

GLM4V inherits from Qwen2.5VL but doesn't need the rope_deltas CFG fix.
Added override to call grandparent method instead of parent to skip
the rope_deltas logic.

notkisk committed:
Override rope_deltas method in GLM4V to prevent inheritance

GLM4V inherits from Qwen2.5VL but doesn't need the rope_deltas CFG fix.
Added override to call grandparent method instead of parent to skip
the rope_deltas logic.

notkisk committed:
Fix rope_deltas corruption in Qwen2.5VL CFG generation and GLM4V inheritance

- Implement proper state management for rope_deltas in Qwen2.5VL models during CFG generation
- Add _update_model_kwargs_for_generation to preserve rope_deltas state between forward passes
- Restore rope_deltas in prepare_inputs_for_generation to ensure correct position calculations
- Override rope_deltas logic in GLM4V to prevent unwanted inheritance from Qwen2.5VL
- Regenerate modular-derived files to maintain consistency

Fixes #39749

notkisk committed:
Fix rope_deltas corruption in Qwen2.5VL CFG generation and GLM4V inheritance

- Implement proper state management for rope_deltas in Qwen2.5VL models during CFG generation
- Add _update_model_kwargs_for_generation to preserve rope_deltas state between forward passes
- Restore rope_deltas in prepare_inputs_for_generation to ensure correct position calculations
- Override rope_deltas logic in GLM4V to prevent unwanted inheritance from Qwen2.5VL
- Regenerate modular-derived files to maintain consistency

Fixes #39749

notkisk committed:
Fix linting issues in GLM4V modular file

- Add missing imports for Any and ModelOutput type hints
- Remove trailing whitespace from blank line
- Maintain GLM4V rope_deltas override functionality

notkisk committed:
Fix linting issues in GLM4V modular file

- Add missing imports for Any and ModelOutput type hints
- Remove trailing whitespace from blank line
- Maintain GLM4V rope_deltas override functionality

notkisk committed:
Apply code formatting to GLM4V modular file

- Run ruff format to ensure consistent code style
- Maintain rope_deltas override functionality
- No functional changes, formatting only

notkisk committed:
Apply code formatting to GLM4V modular file

- Run ruff format to ensure consistent code style
- Maintain rope_deltas override functionality
- No functional changes, formatting only

zucchini-nlp commented on src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:
```diff
@@ -1535,6 +1535,10 @@ def prepare_inputs_for_generation(
     ):
         # Overwritten -- in specific circumstances we don't want to forward image inputs to the model
 
+        # Restore rope_deltas from kwargs if available (needed for CFG generation)
+        if "rope_deltas" in kwargs and kwargs["rope_deltas"] is not None:
+            self.model.rope_deltas = kwargs["rope_deltas"]
+
```
looks quite hacky to me

What if we just pass `rope_deltas` to the model's forward instead of re-assigning the `self` attribute? AFAIR smth similar was done in the first version and it had one test failing. We can just overwrite the test, if it fails still

notkisk replied:
Thanks for the feedback, i revisited it and did a tiny refactor

notkisk committed:
Refactor Qwen2.5VL rope_deltas handling to use parameter-based approach

This commit addresses the maintainer feedback to replace the "hacky" state
mutation approach with a clean parameter-based solution for rope_deltas
during CFG generation.

Changes made:
- Modified forward() to accept rope_deltas as optional parameter
- Use parameter when provided, fall back to model state when not provided
- Only update model state when rope_deltas parameter is not provided
- Updated _update_model_kwargs_for_generation to prioritize outputs.rope_deltas
- Updated prepare_inputs_for_generation to pass rope_deltas as parameter
- Eliminates state corruption during multi-pass CFG generation
- Maintains full backward compatibility with existing code

The fix transforms the approach from mutating model state during forward passes
to a functional parameter-based approach that prevents shared mutable state
issues while preserving all existing functionality.

Fixes rope_deltas corruption in Qwen2.5VL during CFG generation by addressing
the root cause with clean architecture instead of state mutation workarounds.

notkisk committed:
Refactor Qwen2.5VL rope_deltas handling to use parameter-based approach

This commit addresses the maintainer feedback to replace the "hacky" state
mutation approach with a clean parameter-based solution for rope_deltas
during CFG generation.

Changes made:
- Modified forward() to accept rope_deltas as optional parameter
- Use parameter when provided, fall back to model state when not provided
- Only update model state when rope_deltas parameter is not provided
- Updated _update_model_kwargs_for_generation to prioritize outputs.rope_deltas
- Updated prepare_inputs_for_generation to pass rope_deltas as parameter
- Eliminates state corruption during multi-pass CFG generation
- Maintains full backward compatibility with existing code

The fix transforms the approach from mutating model state during forward passes
to a functional parameter-based approach that prevents shared mutable state
issues while preserving all existing functionality.

Fixes rope_deltas corruption in Qwen2.5VL during CFG generation by addressing
the root cause with clean architecture instead of state mutation workarounds.

notkisk committed:
Fix torch.compile compatibility by removing locals() check

The previous implementation used locals() to check if current_rope_deltas
was defined, but torch.compile cannot trace builtin functions like locals().

Changes made:
- Initialize current_rope_deltas at function start to ensure it's always defined
- Remove locals() check from output construction
- Maintain same logical behavior while being compilation-friendly

Fixes torch._dynamo.exc.Unsupported: Failed to trace builtin operator 'locals'
in test_generate_compile_model_forward and test_generate_compilation_all_outputs

notkisk committed:
Fix torch.compile compatibility by removing locals() check

The previous implementation used locals() to check if current_rope_deltas
was defined, but torch.compile cannot trace builtin functions like locals().

Changes made:
- Initialize current_rope_deltas at function start to ensure it's always defined
- Remove locals() check from output construction
- Maintain same logical behavior while being compilation-friendly

Fixes torch._dynamo.exc.Unsupported: Failed to trace builtin operator 'locals'
in test_generate_compile_model_forward and test_generate_compilation_all_outputs

zucchini-nlp commented on src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:
```diff
@@ -1289,21 +1292,25 @@ def forward(
                     attention_mask=attention_mask,
                 )
-                self.rope_deltas = rope_deltas
+                # Only update model state if rope_deltas parameter was not provided
+                if rope_deltas is None:
+                    self.rope_deltas = calculated_rope_deltas
+                current_rope_deltas = calculated_rope_deltas
```
i think we can assign `self.rope_deltas` if we ended up by running this code block, and we don't need to check if users provided `rope_deltas`. The only case when this is possible is pre-fill and we need to re-compute correct `self.rope_deltas` in every prefill stage

zucchini-nlp commented on src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:
```diff
@@ -1585,6 +1595,10 @@ def prepare_inputs_for_generation(
             model_inputs["pixel_values"] = None
             model_inputs["pixel_values_videos"] = None
 
+        # Add rope_deltas parameter for clean CFG generation
+        if rope_deltas_param is not None:
+            model_inputs["rope_deltas"] = rope_deltas_param
+
```
this will be passed over in `super()` call i think, so we don't need to manually pop and add again. Can you check?

zucchini-nlp commented on src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:
```diff
@@ -1730,5 +1744,24 @@ def _expand_dict_for_generation(dict_to_expand):
+            outputs, model_kwargs, is_encoder_decoder, num_new_tokens
+        )
+
+        # Preserve rope_deltas for CFG and multi-pass generation
+        if hasattr(outputs, "rope_deltas") and outputs.rope_deltas is not None:
+            model_kwargs["rope_deltas"] = outputs.rope_deltas
+        elif hasattr(self.model, "rope_deltas") and self.model.rope_deltas is not None:
+            model_kwargs["rope_deltas"] = self.model.rope_deltas
```
checking only output is more reliable, lets' not pass model attributes here

zucchini-nlp reviewed (commented):
Thanks! Can you add the same changes in qwen2-vl and add a small test as well?

notkisk committed:
Fix rope_deltas corruption during CFG generation in Qwen2-VL and Qwen2.5-VL

- Simplify rope_deltas assignment logic in prefill stage
- Remove unnecessary conditional check for rope_deltas
- Update _update_model_kwargs_for_generation method implementation
- Add CFG generation support to Qwen2-VL model
- Add test_rope_deltas_preserved_for_cfg_generation test for both models

Fixes #39749

notkisk committed:
Fix rope_deltas corruption during CFG generation in Qwen2-VL and Qwen2.5-VL

- Simplify rope_deltas assignment logic in prefill stage
- Remove unnecessary conditional check for rope_deltas
- Update _update_model_kwargs_for_generation method implementation
- Add CFG generation support to Qwen2-VL model
- Add test_rope_deltas_preserved_for_cfg_generation test for both models

Fixes #39749

notkisk committed:
Fix code formatting with ruff

- Apply ruff formatting to modular_qwen2_5_vl.py
- Regenerate auto-generated modeling files from modular source

notkisk committed:
Fix code formatting with ruff

- Apply ruff formatting to modular_qwen2_5_vl.py
- Regenerate auto-generated modeling files from modular source

LysandreJik committed:
Fix responses add tests (#39848)

* Quick responses fix

* [serve] Fix responses API and add tests

* Remove typo

* Remove typo

* Tests

yonigozlan committed:
Add fast image processor Janus, Deepseek VL, Deepseek VL hybrid (#39739)

* add fast image processor Janus, deepseek_vl, deepseek_vl_hybrid

* fix after review

sbucaille committed:
[image-processing] deprecate `plot_keypoint_matching`, make `visualize_keypoint_matching` as a standard (#39830)

* fix: deprecate plot_keypoint_matching and make visualize_keypoint_matching for all Keypoint Matching models

* refactor: added copied from

* fix: make style

* fix: repo consistency

* fix: make style

* docs: added missing method in SuperGlue docs

qgallouedec committed:
Allow `TrackioCallback` to work when pynvml is not installed (#39851)

Allow TrackioCallback to work when pynvml is not installed

ArthurZucker committed:
remove dtensors, not explicit (#39840)

* remove dtensors, not explicit

Co-authored-by: 3outeille <3outeille@users.noreply.github.com>

* style

* fix test

* update

* as we broke saving try to fix

* output layouts should exit

* nit

* devicemesh exists if it was distributed

* use _device_mesh of self

* update

* lol

* fix

* nit

* update

* fix!

* this???

* grumble grumble

* ?

* fuck me

---------

Co-authored-by: 3outeille <3outeille@users.noreply.github.com>

qgallouedec committed:
Improve `is_wandb_available` function to verify WandB installation (#39875)

Improve `is_wandb_available` function to verify WandB installation by checking for a key attribute

qgallouedec committed:
Refactor label name handling for PEFT models in Trainer class (#39265)

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

ydshieh committed:
Use comment to build doc on PRs (#39846)

* try

* try

* try

* try

* try

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

zucchini-nlp reviewed (commented):
One tiny last bit, can we update the test to actually run a CFG generation. That way we'll be sure CFG works and doesn't generate gibberish 

You can add a tiny test under `IntegrationTests` with a 3B model

akibjawad committed:
Add support for including in-memory videos (not just files/urls) in apply_chat_template (#39494)

* added code for handling video object ,as dictionary of frames and metadata, in chat template

* added new test where videos are passed as objects (dict of frames, metadata) in the chat template

* modified hardcoded video_len check that does not match with increased number of tests cases.

* Modify hardcoded video_len check that fails with increased number of tests

* update documentation of multi-modal chat templating with extra information about including video object in chat template.

* add array handling in load_video()

* temporary test video inlcuded

* skip testing smolvlm with videos that are list of frames

* update documentation & make fixup

* Address review comments

notkisk committed:
Merge branch 'fix-qwen2-5-vl-rope-deltas-cfg' of https://github.com/notkisk/transformers into fix-qwen2-5-vl-rope-deltas-cfg

notkisk committed:
feat: added real cfg generation tests
